Detectors/TRD/calibration/src/TrackBasedCalib.cxx

```diff
@@ -77,6 +82,85 @@ void TrackBasedCalib::calculateAngResHistos()
   // mAngResHistos.print();
 }
 
+int TrackBasedCalib::filldEdx(gsl::span<const TrackTRD>& tracks, bool isTPCTRD)
+{
+  auto& params = TRDCalibParams::Instance();
+  int nTracksSuccess = 0;
+  for (const auto& trkIn : tracks) {
+
+    if (trkIn.getNtracklets() < params.nTrackletsMinGainCalib) {
+      continue;
+    }
+
+    if (trkIn.getP() < params.pMin || trkIn.getP() > params.pMax) {
+      continue;
+    }
+
+    auto id = trkIn.getRefGlobalTrackId();
+    float dEdxTPC;
+    if (isTPCTRD) {
+      dEdxTPC = mTracksTPC[id].getdEdx().dEdxTotTPC;
+    } else {
+      dEdxTPC = mTracksTPC[mTracksITSTPC[id].getRefTPC()].getdEdx().dEdxTotTPC;
+    }
+    if (dEdxTPC < params.dEdxTPCMin || dEdxTPC > params.dEdxTPCMax) {
+      continue;
+    }
+
+    if (std::isnan(trkIn.getSnp())) {
+      LOG(alarm) << "Track with invalid parameters found: " << trkIn.getRefGlobalTrackId();
+      continue;
+    }
+
+    for (int iLayer = NLAYER - 1; iLayer >= 0; --iLayer) {
+      if (trkIn.getTrackletIndex(iLayer) == -1) {
+        continue;
+      }
+      if (mNoiseCalib && mNoiseCalib->isTrackletFromNoisyMCM(mTrackletsRaw[trkIn.getTrackletIndex(iLayer)])) {
+        // ignore tracklets which originate from noisy MCMs
+        continue;
+      }
+
+      // Remove split tracklets
+      if (trkIn.getIsCrossingNeighbor(iLayer)) {
+        continue;
+      }
+
+      int trkltId = trkIn.getTrackletIndex(iLayer);
+      int trkltDet = mTrackletsRaw[trkltId].getDetector();
+      int trkltSec = trkltDet / (NLAYER * NSTACK);
+
+      const auto& tracklet = mTrackletsRaw[trkltId];
+
+      float Q0 = tracklet.getQ0();
+      float Q1 = tracklet.getQ1();
+      float Q2 = tracklet.getQ2();
+      if (Q0 == 0 || Q1 == 0 || Q2 == 0 || Q0 >= 127 || Q1 >= 127 || Q2 >= 62) {
+        continue;
+      }
+
+      // Correction for tracklet angle
+      const auto& trackletCalib = mTrackletsCalib[trkltId];
+      float tgl = trkIn.getTgl();
+      float snp = trkIn.getSnpAt(o2::math_utils::sector2Angle(trkltSec), trackletCalib.getX(), bz);
+
+      float trkLength = sqrt(1 + snp * snp / (1 - snp * snp) + tgl * tgl);
```

Comment:
can you please add a comment here how you calculate the track length and double check that this is correct?